apps/vscode/src/vdoc/vdoc-tempfile.ts

```diff
@@ -82,19 +87,57 @@ async function deleteDocument(doc: TextDocument) {
   }
 }
 
-// create temp files for vdocs. use a base directory that has a subdirectory
-// for each extension used within the document. this is a no-op if the
-// file already exists
 tmp.setGracefulCleanup();
-const vdocTempDir = tmp.dirSync().name;
-function createVirtualDocTempFile(virtualDoc: VirtualDoc) {
-  const ext = virtualDoc.language.extension;
-  const dir = path.join(vdocTempDir, ext);
-  if (!fs.existsSync(dir)) {
-    fs.mkdirSync(dir);
+const VIRTUAL_DOC_TEMP_DIRECTORY = tmp.dirSync().name;
+
+/**
+ * Creates a virtual document in a temporary directory
+ *
+ * The temporary directory is automatically cleaned up on process exit.
+ *
+ * @param virtualDoc The document to use when populating the temporary file
+ * @returns The path to the temporary file
+ */
+function createVirtualDocTempfile(virtualDoc: VirtualDoc): string {
+  const filepath = generateVirtualDocFilepath(VIRTUAL_DOC_TEMP_DIRECTORY, virtualDoc.language.extension);
+  createVirtualDoc(filepath, virtualDoc.content);
+  return filepath;
+}
+
+/**
+ * Creates a virtual document in the provided directory
+ *
+ * @param virtualDoc The document to use when populating the temporary file
+ * @param directory The directory to create the temporary file in
+ * @returns The path to the temporary file
+ */
+function createVirtualDocLocalFile(virtualDoc: VirtualDoc, directory: string): string {
+  const filepath = generateVirtualDocFilepath(directory, virtualDoc.language.extension);
+  createVirtualDoc(filepath, virtualDoc.content);
+  return filepath;
+}
+
+/**
+ * Creates a file filled with the provided content
+ */
+function createVirtualDoc(filepath: string, content: string): void {
+  const directory = path.dirname(filepath);
+
+  if (!fs.existsSync(directory)) {
+    fs.mkdirSync(directory);
   }
-  const tmpPath = path.join(vdocTempDir, ext, ".intellisense." + uuid.v4() + "." + ext);
-  fs.writeFileSync(tmpPath, virtualDoc.content);
 
-  return tmpPath;
+  fs.writeFileSync(filepath, content);
+}
+
+/**
+ * Generates a unique virtual document file path
+ *
+ * It is important for virtual documents to have unique file paths. If a static
+ * name like `.vdoc.{ext}` is used, it is possible for one language server
+ * request to overwrite the contents of the virtual document while another
+ * language server request is running (#683).
+ */
+function generateVirtualDocFilepath(directory: string, extension: string): string {
+  return path.join(directory, ".vdoc." + uuid.v4() + "." + extension);
```

Comment:
Worth noting that for non-`local` LSP requests this was called `.intellisense.{uuid}.{ext}` but I do not think that name mattered. We now always use `.vdoc.{uuid}.{extension}`.